### PR TITLE
[WIP] Feature/TCP remote atf connection

### PR DESCRIPTION
**Pull request by SNiukalov**
Implement [Remote Automated testing](https://github.com/LuxoftAKutsan/sdl_evolution/blob/remote_atf_proxy/proposals/nnnn-remote_atf_testing.md)

This PR is **not ready** for review.

### Summary
Implement RemoteTestingAdapterServer for QNX x86

##### Breaking Changes
* Added implementation TCP message broker.
* Integration TCP interaction.

### Dependencies:
- [Feature/atf build using cmake](https://github.com/smartdevicelink/sdl_atf/pull/158)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: examples/DynamicSignalsAndSlots.lua [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  print("Slot in qt")
+end
+
+function d.slot_in_qt()
+  print("Slot in qt2")
+end
+
+
+qt.connect (d, "some_signal()", d, "slot_in_qt()")
+d:some_signal()
```
Comment (AByzhynar) [inline examples/DynamicSignalsAndSlots.lua]: @SNiukalov add an empty line at the end

File: modules/SDL.lua [hunk 2]
```diff
@@ -25,47 +29,165 @@ SDL.STOPPED = 0
...
+      }
+    }
+    for _, item in pairs(paramsToUpdate) do
+      SDL.LOGGER.set(item.name, item.value)
+    end
+  end
+end
+
+local function getPathAndName(pPathToFile)
+  local pos = string.find(pPathToFile,"/[^/]*$")
```
Comment (AByzhynar) [inline modules/SDL.lua]: @SNiukalov please add space after comma

File: modules/SDL.lua [hunk 3]
```diff
@@ -25,47 +29,165 @@ SDL.STOPPED = 0
...
+    os.execute( "rm -f " .. pPathToFile)
+  end
+end
+
+local function getParamValue(pContent, pParam)
+  local value
+  for line in pContent:gmatch("[^\r\n]+") do
+    if string.match(line, "^%s*" .. pParam .. "%s*=%s*") ~= nil then
+      if string.find(line, "%s*=%s*$") ~= nil then
+        value = ""
```
Comment (AByzhynar) [inline modules/SDL.lua]: @SNiukalov Can we just return “” instead of break, assignment ...

File: modules/SDL.lua [hunk 4]
```diff
@@ -25,47 +29,165 @@ SDL.STOPPED = 0
...
+  end
+end
+
+--- Structure of SDL build options what to be set
+local function getDefaultBuildOptions()
+  local options = { }
+  options.remoteControl = { sdlBuildParameter = "REMOTE_CONTROL", defaultValue = "ON" }
+  if config.remoteConnection.enabled then
+    options.extendedPolicy = { sdlBuildParameter = "EXTENDED_POLICY", defaultValue = "EXTERNAL_PROPRIETARY" }
+  else
```
Comment (AByzhynar) [inline modules/SDL.lua]: @SNiukalov Do we need HTTP policy flow?

File: modules/SDL.lua [hunk 5]
```diff
@@ -74,15 +196,302 @@ local function setSdlBuildOption(self, optionName, sdlBuildParam, defaultValue)
...
+    local _, e = crts:find("-----END RSA PRIVATE KEY-----", i, true)
+    o.key = crts:sub(s, e) .. "\n"
+    for _, v in pairs({ "crt", "rootCA", "issuingCA" }) do
+      i = e
+      s = crts:find("-----BEGIN CERTIFICATE-----", i, true)
+      _, e = crts:find("-----END CERTIFICATE-----", i, true)
+      o[v] = crts:sub(s, e) .. "\n"
+    end
+    return o
+  end
```
Comment (AByzhynar) [inline modules/SDL.lua]: @SNiukalov add empty lines between local functions please

File: modules/SDL.lua [hunk 6]
```diff
@@ -145,6 +562,47 @@ function SDL:StopSDL()
   sleep(1)
 end
 
+function SDL.ForceStopSDL()
+  if config.remoteConnection.enabled then
+    ATF.remoteUtils.app:StopApp(config.SDL)
+  else
```
Comment (AByzhynar) [inline modules/SDL.lua]: @SNiukalov Can we use more short and simple command [slay](
http://www.qnx.com/developers/docs/6.5.0/index.jsp?topic=%2Fcom.qnx.doc.neutrino_utilities%2Fs%2Fslay.htm) ?

File: modules/config.lua [hunk 7]
```diff
@@ -9,14 +9,62 @@
...
 config.color = true
---- Define URL for HMI connection
-config.hmiUrl = "ws://localhost"
---- Define port for HMI connection
-config.hmiPort = 8087
+
+--- Remote cofiguration
+config.remoteConnection = {}
+config.remoteConnection.enabled = true
+config.remoteConnection.url = "172.31.239.150"
```
Comment (AByzhynar) [inline modules/config.lua]: @SNiukalov is this url unchangeable?

File: modules/config.lua [hunk 8]
```diff
@@ -33,14 +81,18 @@ config.checkAllValidations = false
...
+config.pathToSDLPolicyDB = "/pps/qnx/qdb/config/policy.db"
 --- Define path to SDL interfaces
 -- Example: "/home/user/sdl_panasonic/src/components/interfaces"
 config.pathToSDLInterfaces = ""
 --- Define SDL modification
-config.SDL = "smartDeviceLinkCore"
+config.SDL = "SmartDeviceLink"
 --- Definehost for SDL logs
-config.sdl_logs_host = "localhost"
+config.sdl_logs_host = "172.31.239.150"
```
Comment (AByzhynar) [inline modules/config.lua]: @SNiukalov same question as above...

File: modules/remote/remote_connection.lua [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  setmetatable(res, RemoteConnection.mt)
+  return res
+end
+
+--- Connect with Remote server
+function RemoteConnection.mt.__index:Connect()
+  self.connection = remote:RemoteMqClient(self.url, self.port)
+  if not self.connection then
+      error("RemoteConnection was not established.")
```
Comment (AByzhynar) [inline modules/remote/remote_connection.lua]: @SNiukalov check this file for indents. You use everywhere 2 spaces as indent but here I can see 4

File: modules/remote/remote_constants.lua [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+local RemoteConstants = {}
+
+--- Application status enumeration
+RemoteConstants.APPLICATION_STATUS = {
+  IDLE = -1,
+  NOT_RUNNING = 0,
+  RUNNING = 1,
+}
+--- Error code enumeration
+RemoteConstants.ERROR_CODE = {
```
Comment (AByzhynar) [inline modules/remote/remote_constants.lua]: @SNiukalov align indents

File: modules/remote/remote_file_utils.lua [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+-- *Dependencies:* none
+--
+-- *Globals:* none
+-- @module remote.remote_file_utils
+-- @copyright [Ford Motor Company](https://smartdevicelink.com/partners/ford/) and [SmartDeviceLink Consortium](https://smartdevicelink.com/consortium/)
+-- @license <https://github.com/smartdevicelink/sdl_core/blob/master/LICENSE>
+
+local constants = require("remote/remote_constants")
+
+local RemoteFileUtils = {
```
Comment (AByzhynar) [inline modules/remote/remote_file_utils.lua]: @SNiukalov 6 spaces indent

File: modules/remote/remote_mq_utils.lua [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  end
+
+  if not isResultValid then
+    error("Module RemoteMqUtils received invalid result value: " .. result)
+  end
+
+  if result == constants.ERROR_CODE.SUCCESS then
+    if data == nil then
+      return true
+    else
```
Comment (AByzhynar) [inline modules/remote/remote_mq_utils.lua]: @SNiukalov else is redundant

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,78 @@
+project(ATF)
+cmake_minimum_required(VERSION 3.11)
+include(FetchContent)
+
+# Enable PIC as static libraries may be linked to shared objects
+set (CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+set (CMAKE_CXX_STANDARD 11)
+set (CMAKE_INCLUDE_CURRENT_DIR ON)
+set (CMAKE_AUTOMOC ON)
```
Comment (LuxoftAKutsan) [inline CMakeLists.txt]: @SNiukalov  what is this flag about?

File: modules/applink_connection.lua [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+-- *Dependencies:* `remote/remote_mq_utils`
+--
+-- *Globals:* ATF
+-- @module applinkConnection
+-- @copyright [Ford Motor Company](https://smartdevicelink.com/partners/ford/) and [SmartDeviceLink Consortium](https://smartdevicelink.com/consortium/)
+-- @license <https://github.com/smartdevicelink/sdl_core/blob/master/LICENSE>
+
+local remote_mq_utils = require("remote/remote_mq_utils")
+
+local ApplinkConnection = {
```
Comment (LuxoftAKutsan) [inline modules/applink_connection.lua]: @SNiukalov do not mention applink in open source

File: modules/applink_connection.lua [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  mt = { __index = {} }
+}
+---Appink messege type enumeration
+ApplinkConnection.SDLMsgType ={
+  SDL_MSG_SDL_START = 0x10,
```
Comment (LuxoftAKutsan) [inline modules/applink_connection.lua]: @SNiukalov this is redundant functionality for open source

File: modules/config.lua [hunk 16]
```diff
@@ -9,14 +9,62 @@
...
+config.hmiAdapterConfig.Remote.HMItoSDLMqConfig.max_message_size = 4095
+config.hmiAdapterConfig.Remote.HMItoSDLMqConfig.flags = 257 -- O_WRONLY | O_CREAT
+config.hmiAdapterConfig.Remote.HMItoSDLMqConfig.mode = 1638 -- S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH
+-- Define configuration parameters of SDLtoHMI Mq
+-- for Remote HMI connection
+config.hmiAdapterConfig.Remote.SDLtoHMIMqConfig = {}
+config.hmiAdapterConfig.Remote.SDLtoHMIMqConfig.name = "/FromSDL"
+config.hmiAdapterConfig.Remote.SDLtoHMIMqConfig.max_messages_number = 128
+config.hmiAdapterConfig.Remote.SDLtoHMIMqConfig.max_message_size = 4095
+config.hmiAdapterConfig.Remote.SDLtoHMIMqConfig.flags = 256 -- O_RDONLY | O_CREAT
```
Comment (LuxoftAKutsan) [inline modules/config.lua]: @SNiukalov remove mq staff from open source

**Comment by dboltovskyi** [comment]
Closed in a favor of #180

**Pull request closed without merging**